dialogflow_task_executive/node_scripts/webhook_server.py

```diff
@@ -0,0 +1,123 @@
+        """
+        user_agent = self.headers.get("User-Agent")
+        rospy.loginfo('Recieved POST request' + str(self.headers))
+        if user_agent == "Google-Dialogflow":
+            self._parse_json()
+            self._pub_task()
+            self._response()
+        else:
+            rospy.logwarn('User-Agent header should be Google-Dialogflow, but got ' + user_agent)
+            self._bad_request()
+
+    def _parse_json(self):
+        content_len = int(self.headers.get("content-length"))
+        request_body = self.rfile.read(content_len).decode("utf-8")
+        self.json_content = json.loads(request_body)
+
+    def _pub_task(self):
+        """
+        Publish ROS message to exec task.
+        """
+        msg = DialogResponse()
+        msg.header.stamp = rospy.Time.now()
+        msg.query = self.json_content['queryResult']['queryText']
+        msg.action = self.json_content['queryResult']['action']
+        msg.response = self.json_content['queryResult']['fulfillmentText']
+        if self.json_content['queryResult']['allRequiredParamsPresent'] == 'True':
+            msg.fulfilled = True
+        msg.parameters = json.dumps(self.json_content['queryResult']['parameters'])
+        msg.speech_score = 1.0
+        msg.intent_score = self.json_content['queryResult']['intentDetectionConfidence']
+        rospy.loginfo("The message summary from Dialogflow \n" + str(msg))
+        self.pub.publish(msg)
+
+    def _make_response(self):
+        """
+        This function is in development. If you want to develop the application like the robot sends its state to the Dialogflow and the Dialogflow handle it, you have to make new response_type. 
+        Please see https://cloud.google.com/dialogflow/es/docs/fulfillment-webhook#webhook_response for details.
+        """
+        res_body = {}
+        
+        # default text response
+        # res_body["fulfillmentMessages"] = []
+        # res_body["fulfillmentMessages"].append({})
+        # res_body["fulfillmentMessages"][0]["text"] = {"text": ["Text response from webhook"]}
+        
+        self.res_body = json.dumps(res_body, indent=2)
+        
+    def _response(self):
+        """
+        The DialogFlow client expects the non-empty response.
+        Please see https://cloud.google.com/dialogflow/es/docs/fulfillment-webhook for details.
+        """
+        self._make_response()
+        self.send_response(200)
+        self.send_header('Content-type', 'application/json; charset=utf-8')
+        self.send_header('Content-length', len(self.res_body.encode()))
+        self.end_headers()
+        self.wfile.write(self.res_body.encode('utf-8'))
+
+    def _bad_request(self):
+        self.send_response(400)
+        self.end_headers()
+
+        
+if __name__ == '__main__':
+    try:
+        server = Server()
+        rospy.loginfo('DialogFlow HTTPS Server starts - %s:%s' % (server.host, server.port))
+        server.httpd.serve_forever()
+    except Exception as e:
```

Comment:
why do we need to catch the exception?

Reply:
if you want to catch the exception and try multiple times, please add trial loop